### DISCUSSION TITLE
How are native addons going to interact with the VFS?

### DISCUSSION THREAD

**Discussion by RaisinTen** [nodejs/single-executable#29] [General]
* Is the VFS going to load shared objects in-memory? Is that even possible on all our supported platforms?
* Do we expect native addons to be able to access files inside the VFS? If so, how?

**Comment by jviotti** [comment]
Good questions! 

> Is the VFS going to load shared objects in-memory? Is that even possible on all our supported platforms?

Native add-ons are dynamically shared libraries. I believe you can load them from a random offset of a file on Linux, but they need to be separate files for macOS and Windows. @jesec can correct me if I'm wrong, but I believe PKG extracts them to a temporary directory and loads them from there.

> Do we expect native addons to be able to access files inside the VFS? If so, how?

Hm, I don't think this will be possible at all with the current architecture. We are injecting the VFS logic into Node.js, so if you use any other mechanism of reading files, then they would by definition not work. @jesec does PKG deal with this in any way?

Worst case I guess we can expose the location of the VFS to the add-on and provide some native APIs to read files from there... Of course, native add-ons would have to use that and wouldn't inherit VFS capabilities out of the box.

**Reply by RaisinTen** [reply]
What if we keep the VFS unpacked on disk by the time the single-executable starts running? @arcanis mentioned that even Yarn does a similar thing in https://github.com/nodejs/single-executable/discussions/27#discussioncomment-3465821. I believe doing that would vastly simplify the implementation of the VFS.

**Reply by arcanis** [reply]
> @arcanis mentioned that even Yarn does a similar thing in https://github.com/nodejs/single-executable/discussions/27#discussioncomment-3465821

We only do this for specific packages. For the vast majority, we want to run scripts from the archives without extracting them. How does Electron apps solve this problem? If I remember correctly, Atom used Onigurima 🤔

**Reply by RaisinTen** [reply]
> We only do this for specific packages. For the vast majority, we want to run scripts from the archives without extracting them. 

Why though?

> How does Electron apps solve this problem? If I remember correctly, Atom used Onigurima 🤔

Electron too unpacks files on the disk https://github.com/electron/electron/blob/3de78442127f7a75c2b49c09bccb0277900be28a/shell/common/asar/archive.h#L85-L87
```cc
  // Copy the file into a temporary file, and return the new path.
  // For unpacked file, this method will return its real path.
  bool CopyFileOut(const base::FilePath& path, base::FilePath* out);
```
but I'm not sure if it's done for everything inside the asar or just the native bindings (will have to check)

**Reply by joyeecheung** [reply]
This sounds like JIT on steroid :) On Linux it is possible to load a library from memory (we can just read the original lib and put it into one of our sections when building the SEA, memfd_create() at runtime and dump that lib into the fd returned, then dlopen that fd): https://x-c3ll.github.io/posts/fileless-memfd_create/

There are also utilities for doing that on other platforms like Windows or macOS, not sure how reliable they are though: https://github.com/fancycode/MemoryModule https://github.com/malisal/loaders

**Reply by jviotti** [reply]
Interesting @joyeecheung . We should definitely investigate loading native add-ons from within the VFS then. Sounds like an experiment you can lead @RaisinTen @dsanders11 @robertgzr

**Reply by RaisinTen** [reply]
Yes, we could definitely try experimenting with those but first I wanted to know about the reasons why we don't want to unpack the full VFS.

**Reply by arcanis** [reply]
It's slow, it increases risks of data corruption (especially with concurrent runs), it prevents signing programs, it can't be used on read-only filesystems, it leads to less deterministic execution (case sensitivity). Keeping files together rather than extracted on the filesystem may also lead to [faster execution time](https://www.sqlite.org/fasterthanfs.html) if well implemented.

**Reply by jviotti** [reply]
Yeah, I second what @arcanis said. @RaisinTen another great example of those performance issues due to additional I/O would be anti-viruses kicking in because of those extractions

**Reply by RaisinTen** [reply]
> It's slow, it increases risks of data corruption (especially with concurrent runs)

What if we unpack the SEA just once in some place like the equivalent of the app data directory on all platforms, either during the first run or during the installation phase if we consider that to be a thing?

> it prevents signing programs

Is that true for all applications that unpack things like scripts on the disk? Is there no way to protect such a directory from getting modified by other programs?

> it can't be used on read-only filesystems,

I'm not sure I understand, you would need to have write access to place the binary in the filesystem in the first place, right?

> it leads to less deterministic execution (case sensitivity).

That's not necessarily a bad thing though, the application would work in a way that's native to the file system being used.

FWIW, at Postman we wanted to build something like Tauri but as a native addon, so you would be able to use Node.js in the main process instead of Rust. That would require the native bindings to have access to files like images, videos, etc. Is there no better way to allow this other than creating a new native API for the addons to be able to access files inside the VFS?

**Reply by arcanis** [reply]
> What if we unpack the SEA just once in some place like the equivalent of the app data directory on all platforms, either during the first run or during the installation phase if we consider that to be a thing?

That's pretty much necessary, it'd be far too slow otherwise. It still brings its set of problems: you then need a mutex to prevent multiple processes from corrupting the files (with all problems filesystem mutexes have by themselves), you need to figure out when to purge those files (if ever? even if their binaries are deleted? what if users run multiple versions of the binary?).

> I'm not sure I understand, you would need to have write access to place the binary in the filesystem in the first place, right?

No, you can put the binary when building the image, then make the filesystem read-only for the most part while running the application.

> That's not necessarily a bad thing though, the application would work in a way that's native to the file system being used.

I admit case insensitive filesystems have been a pain every single time I've been confronted to them (hello Git case renames), so I don't see any usefulness to that myself 😅

> Is there no better way to allow this other than creating a new native API for the addons to be able to access files inside the VFS?

Fuse, but that's unlikely to be a reasonable option (non-portable, not-well-maintained on OSX / Windows, perf concerns).

**Reply by jviotti** [reply]
> What if we unpack the SEA just once in some place like the equivalent of the app data directory on all platforms, either during the first run or during the installation phase if we consider that to be a thing?

Runtime I/O is definitely prone to these sort of things. Ideally, we would find a way to not extract anything.

> I'm not sure I understand, you would need to have write access to place the binary in the filesystem in the first place, right?

@RaisinTen This is a very valid use case on enterprise settings. Usually, an IT admin with write access provisions the base system with the applications the employees would use, and then distribute the same thing as a read-only file-system for general consumption.

> Fuse, but that's unlikely to be a reasonable option (non-portable, not-well-maintained on OSX / Windows, perf concerns).

The AppImage project (https://appimage.github.io) is a great example of how to do this well on Linux. AppImages "extract" the app at runtime by mounting it with fuse in a read-only manner. I don't think there are ways to accomplish something like this as well as in Linux in the other operating systems though, but sounds worth exploring. 

@RaisinTen You can try opening a thread on the AppImage project and see if they have any ideas how would AppImages work i.e. on Windows?

**Reply by RaisinTen** [reply]
> Ideally, we would find a way to not extract anything.

What if an SEA wants to run an executable binary? Is there a way to run that as a string without extracting it on the file system?

**Reply by jviotti** [reply]
I believe it can be done on Linux, but not sure about other OSes. We need to do some proper research... We can try to i.e. create a basic macOS terminal C/C++ program, inject another Mach-O executable on it using Postject, and attempt to execute it without extraction.

However, I don't think it will be possible to do for all operating systems. At least we can maybe keep extraction to a minimum?

**Reply by RaisinTen** [reply]
> I believe it can be done on Linux, but not sure about other OSes.

Doing it on linux too seems to be complicated according to the discussion in https://stackoverflow.com/a/66053812.

> At least we can maybe keep extraction to a minimum?

That seems to be the most viable option.

Another observation - even if we manage to run executables in memory  / load shared objects in memory, those files might depend on files present in certain paths on the real file system. It won't be possible to configure those binaries to use the VFS C++ API if the user doesn't have access to the source code of those binaries.

**Reply by jviotti** [reply]
Hm, not sure what the best solution would be here then. The only model I know that solves this problem is the FUSE mounting approach followed by AppImages. FUSE on Linux is super fast, so you don't encounter all the issues @arcanis raised, but I don't know if it's possible in other OSes at all.

 Can you research whether it's feasible to do FUSE in other systems, or at least use something that is relatively close? For example, DMGs on macOS come pretty close in that you don't "extract" the DMG when double-clicking, but only mount a "view" into the DMG without occupying extra space, but DMG mounting is pretty slow as far as I understand.

We could try:

- Not signing DMGs to avoid the mounting integrity checks
- Doing DMG without any compression at all and see how that would go

However, FUSE can ensure that the file system is always unmounted post-execution, which is not the case with DMGs...

**Comment by jviotti** [comment]
As per the call today, let's park this thread for the time-being. We'll focus on making VFS implementations possible from user-land, and we can revisit the problem of a blessed VFS implementations later on. The reason for this strategic decision is that designing a truly universal VFS seems be a very complex problem. We can focus on the other parts of SEAs for now instead of being blocked on VFS for a long time.
